src/struct_mv/coarsen.c

```diff
@@ -478,6 +478,10 @@ hypre_StructCoarsen( hypre_StructGrid  *fgrid,
    /* create new coarse grid */
    hypre_StructGridCreate(comm, ndim, &cgrid);
 
+   /* Set global size to a number different than zero to
+      avoid its computation on hypre_StructGridAssemble. */
+   hypre_StructGridGlobalSize(cgrid) = -1;
```

Comment:
Where will this be set to a correct value?

Reply:
Hi @rfalgout, it will be computed at `hypre_StructMatmultCommunicate`: https://github.com/hypre-space/hypre/blob/c7cab52acc7e8a20c0246f546133ff2c0a190a10/src/struct_mv/struct_matmult.c#L878

I believe it's a bit counter-intuitive to have a data member from `hypre_StructGrid` being computed at a `hypre_StructMatrix` function, but this solution was the best that I could find to reduce the number of `MPI_Allreduce` calls. What do you think about this?

Reply:
The problem is that Coarsen should return a valid Grid.  Maybe we can redefine Coarsen to return an unassembled Grid, which would require us to explicitly call Assemble.  I don't know if that would help here, because I don't really understand how this reduces calls to Allreduce to begin with.  Can you explain in words or pseudocode?  I'm being lazy. :)

Reply:
I think it's a good idea to change the behavior of `hypre_StructCoarsen` to return an unassembled grid. 

Currently, `hypre_StructCoarsen` calls `hypre_StructGridAssemble`, which has four `MPI_Allreduce` calls (one from that function itself and the other 3 from `hypre_BoxManAssemble` that is called from GridAssemble). In the semi-struct matrix multiplication setup, we have as many calls to `hypre_StructCoarsen` as the number of parts multiplied by the number of variables. So, if we return an unassembled grid from `hypre_StructCoarsen`, we could avoid `nparts * nvars` calls to MPI_Allreduce to a single one with vector size `nparts * nvars` times larger than the respective one in the old code. Also, this would solve the blocking behavior in `SStructMatmultSetup` due to the collective call (allreduce). 

I'm not sure if that was sufficiently clear, we can meet otherwise :)

Reply:
PS: although the new interface for semi-struct matrix multiplication in this PR showed speedup wrt to the old, there's still some serialization happening in the setup function for semi-struct matrix multiplication (`hypre_SStructMatrixMultSetup`).

Reply:
This definitely helps, Victor.  Thanks!  One more question.  I had misunderstood you to say that the global sizes for each of StructGrid grids were eventually computed, but that doesn't appear to be the case.  Is it really a good idea to have incomplete objects here?  Maybe we should discuss some more redesign ideas here.  For example, if I recall correctly, the SStructMatrix keeps a number of "variable grids" around that are reused in the different StructMatrix matrices to avoid replicating a bunch of grid objects (so each of the SStructMatrix matrices have references to possibly the same grids).  Are you doing anything like that here?  If not, maybe we can think about making some additional changes to the Matmult interface to help us have more control over grid construction in the SStruct code.

Actually, now that I look at this a bit closer, I wonder if we should maybe just remove the GlobalSize component of the grid altogether.  It doesn't look like it serves any useful purpose other than to help us compute flop information.  What do you think?  Or we could just turn it on when we have logging turned up high enough or something?

Are you planning to make the "Matmult" name change for SStructMatrix and SStructPMatrix as well (i.e., SStructMatrixMult -> SStructMatmult and SStructPMatrixMult -> SStructPMatmult)?  We should.

Reply:
Hi @rfalgout,

> I had misunderstood you to say that the global sizes for each of StructGrid grids were eventually computed, but that doesn't appear to be the case

In the case of a `hypre_StructGrid` generated from `hypre_SStructMatrixMultSetup`, its global size will be computed at https://github.com/hypre-space/hypre/blob/c7cab52acc7e8a20c0246f546133ff2c0a190a10/src/sstruct_mv/sstruct_grid.c#L474

> Is it really a good idea to have incomplete objects here? Maybe we should discuss some more redesign ideas here

I don't feel very comfortable with having incomplete objects either, but this was the best solution I could find to avoid having multiple calls to `MPI_Allreduce` from each of the struct grids belonging to a `hypre_SStructGrid`. Setting `global_size = -1` bypasses the computation of  `global_size` in `hypre_StructGridAssemble` and delays such calculation to `hypre_SStructGridComputeGlobalSize`, which has a single `MPI_Allreduce` that computes the global sizes of all `nparts*nvars` struct grids. Would you have some other idea in mind?

> Are you doing anything like that here?

If I understand the question correctly, the answer is yes. I do reuse the struct grids pointed by the `sgrids` member of a `hypre_SStructPGrid` in a `hypre_SStructMatrix` and its belonging struct matrices.

> It doesn't look like it serves any useful purpose other than to help us compute flop information. What do you think?

Good point, `global_size` is mostly used to compute flops. A more important use case though is on `hypre_PFMGComputeDxyz` and also in the SSAMG counterpart:
https://github.com/hypre-space/hypre/blob/c7cab52acc7e8a20c0246f546133ff2c0a190a10/src/struct_ls/pfmg_setup.c#L681

This is performed only on the fine grid matrix though.

> Are you planning to make the "Matmult" name change for SStructMatrix and SStructPMatrix as well

Yes,  I am planning. I started changed some names and pushed the changes, but haven't finished all the work yet :)

Reply:
This all sounds good.  Regarding `global_size` and the use case you point out, I'm not sure we really need it there anyway.  I can take a look at some point to see, but I think the way you are handling this now is okay for the time being.  Let's keep in mind trying to remove this value altogether, though.  BTW, I think some of the logic in the Dxyz routines needs to be updated to match a recent corner case I fixed on master (unless you've already done that).  I also added a comment there that will help us to decode this routine more easily in the future. :)  Thanks again!

Reply:
Ok, thanks Rob! I updated the Dxyz routines in this branch according to master.

Reply:
Awesome!